p2p/rating/peersRatingHandler.go

```diff
@@ -0,0 +1,251 @@
+		return
+	}
+
+	newRating := oldRating + increaseFactor
+	if newRating > maxRating {
+		return
+	}
+
+	prh.updateRating(pid, oldRating, newRating)
+}
+
+// DecreaseRating decreases the rating of a peer with the decrease factor
+func (prh *peersRatingHandler) DecreaseRating(pid core.PeerID) {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	oldRating, found := prh.getOldRating(pid)
+	if !found {
+		// new pid, add it with default rating
+		prh.topRatedCache.Put(pid.Bytes(), defaultRating, int32Size)
+		return
+	}
+
+	newRating := oldRating + decreaseFactor
+	if newRating < minRating {
+		return
+	}
+
+	prh.updateRating(pid, oldRating, newRating)
+}
+
+func (prh *peersRatingHandler) getOldRating(pid core.PeerID) (int32, bool) {
+	oldRating, found := prh.topRatedCache.Get(pid.Bytes())
+	if found {
+		oldRatingInt, _ := oldRating.(int32)
+		return oldRatingInt, found
+	}
+
+	oldRating, found = prh.badRatedCache.Get(pid.Bytes())
+	if found {
+		oldRatingInt, _ := oldRating.(int32)
+		return oldRatingInt, found
+	}
+
+	return defaultRating, found
+}
+
+func (prh *peersRatingHandler) updateRating(pid core.PeerID, oldRating, newRating int32) {
+	oldTier := computeRatingTier(oldRating)
+	newTier := computeRatingTier(newRating)
+	if newTier == oldTier {
+		if newTier == topRatedTier {
+			prh.topRatedCache.Put(pid.Bytes(), newRating, int32Size)
+		} else {
+			prh.badRatedCache.Put(pid.Bytes(), newRating, int32Size)
+		}
+
+		return
+	}
+
+	prh.movePeerToNewTier(newRating, pid)
+}
+
+func computeRatingTier(peerRating int32) string {
+	if peerRating >= defaultRating {
+		return topRatedTier
+	}
+
+	return badRatedTier
+}
+
+func (prh *peersRatingHandler) movePeerToNewTier(newRating int32, pid core.PeerID) {
+	newTier := computeRatingTier(newRating)
+	if newTier == topRatedTier {
+		prh.badRatedCache.Remove(pid.Bytes())
+		prh.topRatedCache.Put(pid.Bytes(), newRating, int32Size)
+	} else {
+		prh.topRatedCache.Remove(pid.Bytes())
+		prh.badRatedCache.Put(pid.Bytes(), newRating, int32Size)
+	}
+}
+
+// GetTopRatedPeersFromList returns a list of random peers, searching them in the order of rating tiers
+func (prh *peersRatingHandler) GetTopRatedPeersFromList(peers []core.PeerID, numOfPeers int) []core.PeerID {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	isListEmpty := len(peers) == 0
+	if numOfPeers < minNumOfPeers || isListEmpty {
+		return make([]core.PeerID, 0)
+	}
+
+	if prh.hasEnoughTopRated(peers, numOfPeers) {
+		return prh.extractRandomPeers(prh.topRatedCache.Keys(), numOfPeers)
+	}
+
+	peersForExtraction := make([][]byte, 0)
+	peersForExtraction = append(peersForExtraction, prh.topRatedCache.Keys()...)
+	peersForExtraction = append(peersForExtraction, prh.badRatedCache.Keys()...)
+
+	return prh.extractRandomPeers(peersForExtraction, numOfPeers)
+}
+
+func (prh *peersRatingHandler) hasEnoughTopRated(peers []core.PeerID, numOfPeers int) bool {
+	counter := 0
+
+	for _, peer := range peers {
```

Comment:
instead of this for loop, can we simply use:
```GO
return len(prh.topRatedCache.Keys()) >= numOfPeers
```

Reply:
refactored this section to split the received peers into tiers based on their ratings